### ISSUE TITLE
Add a `.list.value_counts()` expression

### ISSUE THREAD

**Issue by MisterKloudy** [Eventual-Inc/Daft#2862]
Would return the counts of each element in the lists like the pandas .value_counts() or numpy .unique(with_counts=True) functionality.

**Example:**
```
df = daft.from_pydict({"a": [[1, 2, 2, 3, 3, 3], [1, 3, 5, 5]]})
df.with_column("b", col("a").list.value_counts()).select("b").show()
```

**Expected output:**
```
╭─────────────╮
│ b           │ 
│ ---         │
│ List[Struct[value: Utf8, count: Int32]] │
╞═════════════╡
│ [{value: 1, count: 1},{value: 2, count: 2},{value: 3, count: 3}]   │
├╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ [{value: 1, count: 1},{value: 3, count: 1},{value: 5, count: 2}]   │
╰─────────────╯

(Showing first 2 of 2 rows)
```

**Comment by jaychia** [comment]
Hi @MisterKloudy !

Would the following be possible instead?

```python
df = daft.from_pydict({"id": [1, 2], "a": [[1, 2, 2, 3, 3, 3], [1, 3, 5, 5]]})
df = df.explode("a")
df = df.groupby("id", "a").agg(df["a"].count().alias("value_count"))
```

Output:

```
In [7]: df.collect()
Out[7]:                                                                                                                                          
╭───────┬───────┬─────────────╮
│ id    ┆ a     ┆ value_count │
│ ---   ┆ ---   ┆ ---         │
│ Int64 ┆ Int64 ┆ UInt64      │
╞═══════╪═══════╪═════════════╡
│ 1     ┆ 3     ┆ 3           │
├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2     ┆ 1     ┆ 1           │
├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1     ┆ 2     ┆ 2           │
├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1     ┆ 1     ┆ 1           │
├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2     ┆ 3     ┆ 1           │
├╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2     ┆ 5     ┆ 2           │
╰───────┴───────┴─────────────╯

(Showing first 6 of 6 rows)
```

If you wanted it in exactly the format you just mentioned:

```
# Groups the columns into a struct
df = df.select(
    "id",
    daft.to_struct("a", "value_count").alias("value_count")
)
# Groupby the ID and then aggregate the struct column into a list of structs
df = df.groupby("id").agg(daft.col("value_count").agg_list())
```

It appears that there are some TODO implementations to make `daft.col("value_count").agg_list()` work though. Let me knock out a quick PR here.

**Comment by MisterKloudy** [comment]
@jaychia Thanks for this! 

If my dataframe contains another data-heavy column which is not part of the expression but part of the df, such as a 'text' colum, and col "a" is the list of words from the text column, would the explode expression cause the entire text column to be repeated by the length of col "a" in memory or will the expression be handled independently from the rest of the dataframe? I also wanted to use with_column to avoid having to join the results back by id after the groupby. Would this be a valid concern?

**Comment by jaychia** [comment]
Valid concerns!

I actually talked briefly with the team. We should probably just add a .count_values like you said and return a Map type column

The Map would have keys of the same type as the list's child, and values of int32

That would make it really easy then to retrieve the count per key!

--
Jay Chia
Founder, Eventual ( https://www.eventualcomputing.com/ )

**Comment by MisterKloudy** [comment]
That would be great! The next thing I wanted to do was to retrieve the value with the highest count for each row. At first I was thinking of implementing a UDF to retrieve the list of structs, implement a custom sorting order, then return the first value. Any idea on how this could be optimized?

**Comment by jaychia** [comment]
Implemented in: #2902
